### PR TITLE
optimize: fix eslint in `@base/object` module

### PR DESCRIPTION
**Pull request by PPpro**
Re: https://github.com/cocos/cocos-engine/issues/16375

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/core/data/class.ts [hunk 1]
```diff
@@ -141,36 +141,36 @@ function getDefault (defaultVal): any {
...
+        (ctor as any).$super = baseClass;
     }
 
-    js.setClassName(className, ctor);
+    js.setClassName(className!, ctor);
     return ctor;
 }
 
-function define (className, baseClass, options): any {
+function define (className: string | undefined, baseClass: Constructor | null, options: Parameters<typeof CCClass>[0]): Constructor {
```
Comment (minggo) [inline cocos/core/data/class.ts:166]: Why class name can be `undefined`?

Comment (PPpro) [reply cocos/core/data/class.ts:166]: It's not a good interface design, but as it is used in `CCClass` method, it should be typeof `string | undefined`

https://github.com/cocos/cocos-engine/blob/c9d2f14086f45b18e0eccafdaf8b93d1b4200017/cocos/core/data/class.ts#L286-L297

Comment (minggo) [reply cocos/core/data/class.ts:166]: Can it work if passing undefined?

Comment (PPpro) [reply cocos/core/data/class.ts:166]: yes, actually in the past, `undefined` was often passed in

Comment (PPpro) [reply cocos/core/data/class.ts:166]: I think this line fix the `undefined` issue
https://github.com/cocos/cocos-engine/blob/c9d2f14086f45b18e0eccafdaf8b93d1b4200017/cocos/core/data/class.ts#L178


There is much HACK code in this module :(
I try to add more type detail for it

**Review by minggo — APPROVED** [review]

**Pull request merged**